### ISSUE TITLE
Use cProfile for twistd profiling on 2.5

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2469]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#2469|
|Type|enhancement|
|Created|2007-02-22 15:25:03Z|
|Branch|https://github.com/twisted/twisted/tree/cprofile-2469|

Attachments:

* [cprofile.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/46b/46b97e6934ff6cda7bf3475ae9287ca1602723a8/91d3d0d0de6723a58ecc0d26b03a824ea6c305ac.diff) (583 bytes) - added by jack on 2007-03-02 17:16:35Z - patch to add cprofile support (falls back to profile for older pythons)

<details><summary>Searchable metadata</summary>

```
trac-id__2469 2469
type__enhancement enhancement
reporter__itamarst itamarst
priority__highest highest
milestone__ 
branch__branches_cprofile_2469 branches/cprofile-2469
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1172157903000000 1172157903000000
changetime__1203453811000000 1203453811000000
version__None None
owner__ 
cc__jack cc__twonds cc__zooko cc__dreid
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_4"></a>|@zooko commented|
|-|-|

I just looked at the attached diff and it looks right.  What's the next step to push this into trunk?

In a related issue:

http://docs.python.org/lib/module-hotshot.html

^-- this says "For common usages it is recommended to use cProfile instead. hotshot is not maintained and might be removed from the standard library in the future."

So perhaps the logic of using hotshot by default and falling back to profile/cProfile should be reversed to use profile/cProfile by default.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

The problem is that currently twistd is very hard to test (because of a structure problem), and no feature will be added before we manage to test it.

The ticket with the highest priority on this subject is [#2571](https://github.com/twisted/twisted/issues/2571), but I'm unable to progress on it. Any help here is welcome :).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [22523]) Branching to 'cprofile-2469'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [22524]) Add cprofile runner, with tests.

Refs [#2469](https://github.com/twisted/twisted/issues/2469)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

`twisted.test.test_twistd.AppProfilingTestCase.test_cProfile` and `twisted.test.test_twistd.AppProfilingTestCase.test_cProfileSaveStats` fail on the dumb debian pstats-less configuration.  Guess they should check pstats availability and skip if it is not around.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

Thanks for the check, it looks better.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_11"></a>|@dreid set owner to @therve|
|-|-|

Great work as always therve, however my gut reaction is that the AppProfiler.profilers dictionary should be keyed on the module name.  i.e. profile, hotshot, and cProfile with an uppercase P.

In fact we call self._reportImportError("cProfile", e) when "cprofile" is specified on the command line, so I definitely think this needs to be addressed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

(In [22621]) Rename option flag to cProfile.

Refs [#2469](https://github.com/twisted/twisted/issues/2469)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_15"></a>|@dreid set owner to @therve|
|-|-|

Thanks therve, this looks great, merge away.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set status to `closed`|
|-|-|

(In [22627]) Merge cprofile-2469

Author: therve
Reviewers: dreid, exarkun
Fixes [#2469](https://github.com/twisted/twisted/issues/2469)

Add cProfile to the list of available profilers for twistd.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
